### PR TITLE
[BYOC][ACL] Support add operation

### PR DESCRIPTION
**Pull request by d-smirnov**
Added support for an "add" operation implemented via ACL + unit test

### CODE REVIEW

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 1]
```diff
@@ -161,6 +161,17 @@ def l2_pool2d_pattern():
         pattern = is_op("sqrt")(pattern)
         return pattern
 
+    def add_pattern():
+        """Create an add pattern.
+
+        Returns
+        -------
+        pattern : dataflow_pattern.AltPattern
+            Denotes the convolution pattern.
```
Comment (comaniac) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:170]:
```suggestion
            Denotes the add pattern.
```

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 2]
```diff
@@ -161,6 +161,17 @@ def l2_pool2d_pattern():
...
+        """
+        pattern = is_op("add")(wildcard(), is_constant(), is_constant())
```
Comment (comaniac) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:172]: Why add has 3 arguments? Did I miss something?

File: tests/python/contrib/test_arm_compute_lib/test_add.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            {
+                "lhs_scale": relay.const(0.0126863, "float32"),
+                "lhs_zero_point": relay.const(127, "int32"),
+                "rhs_scale": relay.const(0.0126863, "float32"),
+                "rhs_zero_point": relay.const(127, "int32"),
+                "output_scale": relay.const(0.0126863, "float32"),
+                "output_zero_point": relay.const(127, "int32"),
+            },
+        ),
+    ]:
```
Comment (comaniac) [inline tests/python/contrib/test_arm_compute_lib/test_add.py:104]: 1. This is hard to follow. If possible, could you make a list of 3 elements and iterate the list here?
2. Could you elaborate the purpose of testing 2 QNN adds?

File: tests/python/contrib/test_arm_compute_lib/test_add.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            "dtype": [[dtype]],
+        },
+    }
+
+    input_a = {"op": "input", "name": "", "attrs": {"shape": [[list(shape)]], "dtype": [[dtype]]}}
+    input_b = {"op": "input", "name": "", "attrs": {"shape": [[list(shape)]], "dtype": [[dtype]]}}
+    return [input_a, input_b, node]
+
+
+def test_add():
```
Comment (comaniac) [inline tests/python/contrib/test_arm_compute_lib/test_add.py:59]: Better to be "test_runtime_add".

File: tests/python/contrib/test_arm_compute_lib/test_add.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+
+def test_codegen_add():
+    if skip_codegen_test():
+        return
+
+    shape = (1, 1, 1, 1000)
+    inputs = {"a", "b"}
+    for dtype in ["float32", "uint8"]:
+        for new_shape in [(1, 1000), (10, 10, 10)]:
```
Comment (comaniac) [inline tests/python/contrib/test_arm_compute_lib/test_add.py:134]: `new_shape` is never used...

**Review by comaniac — CHANGES_REQUESTED** [review]

File: src/runtime/contrib/arm_compute_lib/acl_runtime.cc [hunk 6]
```diff
@@ -401,6 +406,50 @@ class ACLRuntime : public JSONRuntimeBase {
...
+     * @param[in]  input2 Second tensor input. Data types supported: U8/QASYMM8/S16/F16/F32
+     * @param[out] output Output tensor. Data types supported: U8/QASYMM8/S16/F16/F32
+     * @param[in]  policy Policy to use to handle overflow.
+     * void configure(ITensor *input1, ITensor *input2, ITensor *output, ConvertPolicy policy);
+     */
+    auto f = std::make_shared<arm_compute::NEArithmeticAddition>();
+    f->configure(&layer->inputs[0], &layer->inputs[1], &layer->outputs[0],
+                 arm_compute::ConvertPolicy::WRAP);
+    layer->function = f;
+  }
```
Comment (lhutton1) [inline src/runtime/contrib/arm_compute_lib/acl_runtime.cc:451]: I think it would be simpler to cover qnn.add in the same function as add, a bit like the CreateConvolution2DLayer function

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 2 above)
Comment (lhutton1) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:172]: I'm not sure we need to define a pattern for add? as it is a single op we shouldn't need to wrap it within a composite function

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 7]
```diff
@@ -337,3 +354,17 @@ def global_avg_pool2d(attrs, args):
...
         return False
     return True
+
+
+@tvm.ir.register_op_attr("add", "target.arm_compute_lib")
+@tvm.ir.register_op_attr("qnn.add", "target.arm_compute_lib")
+def add(attrs, args):
+    """Check if the external ACL codegen for add should be used."""
+    for typ in [args[0].checked_type, args[1].checked_type]:
+        if typ.dtype not in ["float32", "uint8"]:
```
Comment (lhutton1) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:364]: I think it would be better to write add and qnn.add separately. This way we can state add must have float32 input tensors and qnn.add must have uint8 inputs. Currently we allow a mixture of types.

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 4 above)
Comment (lhutton1) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:59]: Probably best to do this for the rest of the tests also

File: tests/python/contrib/test_arm_compute_lib/test_add.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                outputs.append(build_and_run(func, inputs, 1, None, device, enable_acl=acl)[0])
+
+            config = {
+                "shape": shape,
+                "dtype": dtype,
+                "inputs": inputs,
+                "operation": op,
+                "op_params": op_params,
+            }
+            verify(outputs, atol=atol, rtol=rtol, config=config)
```
Comment (lhutton1) [inline tests/python/contrib/test_arm_compute_lib/test_add.py:124]: When we test qnn.add we should also use the parameter `verify_saturation=True`

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 3 above)
Comment (lhutton1) [inline tests/python/contrib/test_arm_compute_lib/test_add.py:104]: Why not test `add`?

File: tests/python/contrib/test_arm_compute_lib/test_add.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+def test_codegen_add():
+    if skip_codegen_test():
+        return
+
+    shape = (1, 1, 1, 1000)
+    inputs = {"a", "b"}
+    for dtype in ["float32", "uint8"]:
+        for new_shape in [(1, 1000), (10, 10, 10)]:
+            func = _get_model(shape, dtype, iter(inputs), relay.add, {})
```
Comment (lhutton1) [inline tests/python/contrib/test_arm_compute_lib/test_add.py:135]: Why not test `qnn.add`?

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 1 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:170]: removed

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 2 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:172]: removed

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 7 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:364]: 1. Yep. The pattern is not needed. Removed. 
2. Yep. Has been split

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 9 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:135]: Added

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 8 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:124]: Could you elaborate the assertion of "0.25 * outs[0].asnumpy().size" for small sized (e.g. (2,2) or similar ) test cases, please?

Comment (lhutton1) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:124]: The intention was to ensure that most of the values output are not saturated i.e. most values are not 255 or 0. I can see why this would cause an issue with test cases of 2x2 as a single value that is 255 or 0 would trigger the asserts. I think the best approach here would be to increase the size of the test case?

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 3 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:104]: 1. Done.
2. Removed

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 4 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:59]: added "runtime"

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 5 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:134]: fixed

File: src/runtime/contrib/arm_compute_lib/acl_runtime.cc (hunk 6 above)
Comment (d-smirnov) [reply src/runtime/contrib/arm_compute_lib/acl_runtime.cc:451]: updated

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 8 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:124]: add_QASYMM8_QASYMM8_QASYMM8 currently always saturate result. So verify_saturation=True cannot be used

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 10]
```diff
@@ -345,3 +345,23 @@ def maximum(attrs, args):
     type_a = args[0].checked_type
     type_b = args[0].checked_type
     return (type_a.dtype == "float32") and (type_b.dtype == "float32")
+
+
+@tvm.ir.register_op_attr("add", "target.arm_compute_lib")
+def add(attrs, args):
+    """Check if the external ACL codegen for add should be used."""
+    for typ in [args[0].checked_type, args[1].checked_type]:
+        if typ.dtype not in ["float32"]:
```
Comment (comaniac) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:354]:
```suggestion
        if typ.dtype != "float32":
```

File: python/tvm/relay/op/contrib/arm_compute_lib.py [hunk 11]
```diff
@@ -345,3 +345,23 @@ def maximum(attrs, args):
...
+            return False
+
+    return True
+
+
+@tvm.ir.register_op_attr("qnn.add", "target.arm_compute_lib")
+def qnn_add(attrs, args):
+    """Check if the external ACL codegen for add should be used."""
+    for typ in [args[0].checked_type, args[1].checked_type]:
+        if typ.dtype not in ["uint8"]:
```
Comment (comaniac) [inline python/tvm/relay/op/contrib/arm_compute_lib.py:364]:
```suggestion
        if typ.dtype != "uint8":
```

File: src/runtime/contrib/arm_compute_lib/acl_runtime.cc [hunk 12]
```diff
@@ -140,8 +141,13 @@ class ACLRuntime : public JSONRuntimeBase {
           CreateGlobalPoolingLayer(&layer_, node);
         } else if ("reshape" == op_name) {
           CreateReshapeLayer(&layer_, node);
+<<<<<<< HEAD
         } else if ("maximum" == op_name) {
           CreateMaximumLayer(&layer_, node);
+=======
+        } else if ("add" == op_name || "qnn.add" == op_name) {
+          CreateAddLayer(&layer_, node);
+>>>>>>> a7fa43daf... ACL: add operation
```
Comment (comaniac) [inline src/runtime/contrib/arm_compute_lib/acl_runtime.cc:150]: Fix conflict.

File: src/runtime/contrib/arm_compute_lib/acl_runtime.cc [hunk 13]
```diff
@@ -416,15 +422,54 @@ class ACLRuntime : public JSONRuntimeBase {
...
+        layer->outputs.push_back(MakeACLTensorFromJSONNode(node));
+      } else if ("qnn.add" == op_name) {
+        layer->inputs.push_back(MakeACLTensorFromJSONEntry(
+            node.GetInputs()[0], &node.GetInputs()[2], &node.GetInputs()[3]));
+        layer->inputs.push_back(MakeACLTensorFromJSONEntry(
+            node.GetInputs()[1], &node.GetInputs()[4], &node.GetInputs()[5]));
+        layer->outputs.push_back(
+            MakeACLTensorFromJSONNode(node, &node.GetInputs()[6], &node.GetInputs()[7]));
+      } else {
+        LOG(FATAL) << "Unsupported op: " << op_name;
```
Comment (comaniac) [inline src/runtime/contrib/arm_compute_lib/acl_runtime.cc:446]: IMO here can simply throw out assertion, since you have checked if `op_name` is `add` or `qnn.add` before invoking this function in `BuildEngine`.

**Review by comaniac — CHANGES_REQUESTED** [review]

File: src/runtime/contrib/arm_compute_lib/acl_runtime.cc (hunk 12 above)
Comment (d-smirnov) [reply src/runtime/contrib/arm_compute_lib/acl_runtime.cc:150]: Done

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 11 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:364]: done

File: python/tvm/relay/op/contrib/arm_compute_lib.py (hunk 10 above)
Comment (d-smirnov) [reply python/tvm/relay/op/contrib/arm_compute_lib.py:354]: done

File: src/runtime/contrib/arm_compute_lib/acl_runtime.cc (hunk 13 above)
Comment (d-smirnov) [reply src/runtime/contrib/arm_compute_lib/acl_runtime.cc]: replaced with throw

**Review by comaniac — APPROVED** [review]
LGTM. Please fix the CI errors.

File: tests/python/contrib/test_arm_compute_lib/test_add.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+                "operation": op,
+                "op_params": op_params,
+            }
+
+            # verify_saturation=False as the result of add_QASYMM8_QASYMM8_QASYMM8
+            # is always saturated currently.
+            verify(outputs, atol=atol, rtol=rtol, config=config, verify_saturation=False)
+
+
+def test_runtime_codegen_add():
```
Comment (lhutton1) [inline tests/python/contrib/test_arm_compute_lib/test_add.py:118]: To avoid confusion I think we should keep codegen tests named `test_codegen_add`

File: tests/python/contrib/test_arm_compute_lib/test_add.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+            config = {
+                "shape": shape,
+                "dtype": dtype,
+                "inputs": inputs,
+                "operation": op,
+                "op_params": op_params,
+            }
+
+            # verify_saturation=False as the result of add_QASYMM8_QASYMM8_QASYMM8
+            # is always saturated currently.
```
Comment (lhutton1) [inline tests/python/contrib/test_arm_compute_lib/test_add.py:114]: Could this be because of the qnn params being used in the test? We can generate qnn params that won't saturate by looking at this example https://github.com/apache/incubator-tvm/blob/master/tests/python/contrib/test_ethosn/test_addition.py#L45. I think we should try and fix this as we won't checking the result is accurate if it is saturated.

If this can be enabled for `qnn.add`, then we will need to make sure it's disabled for `add`.

File: src/runtime/contrib/arm_compute_lib/acl_runtime.cc [hunk 16]
```diff
@@ -417,6 +420,45 @@ class ACLRuntime : public JSONRuntimeBase {
...
+                                                         &node.GetInputs()[3]));
+      layer->inputs.push_back(MakeACLTensorFromJSONEntry(node.GetInputs()[1], &node.GetInputs()[4],
+                                                         &node.GetInputs()[5]));
+      layer->outputs.push_back(
+          MakeACLTensorFromJSONNode(node, &node.GetInputs()[6], &node.GetInputs()[7]));
+    } else {
+      throw std::runtime_error("Unsupported form of add op: " + op_name);
+    }
+
+    /** Initialise the kernel's inputs, output and conversion policy.
```
Comment (lhutton1) [inline src/runtime/contrib/arm_compute_lib/acl_runtime.cc:446]: Personally I don't think we need this doc string, feel free to ignore though

**Review by lhutton1 — COMMENTED** [review]
Almost there :) just a couple of small things and a concern about a saturated test.

**Comment by comaniac** [comment]
@d-smirnov would you address the comments from @lhutton1 or?

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 15 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:114]: ACL's _add_QASYMM8_QASYMM8_QASYMM8_ implemented in a way that it always saturates result. It just cannot be verified as uint8 always cast to QASYMM8. As for the parameter combination which does not cause saturation for qnn.add could you elaborate why this is needed?

File: src/runtime/contrib/arm_compute_lib/acl_runtime.cc (hunk 16 above)
Comment (d-smirnov) [reply src/runtime/contrib/arm_compute_lib/acl_runtime.cc:446]: removed

File: tests/python/contrib/test_arm_compute_lib/test_add.py (hunk 14 above)
Comment (d-smirnov) [reply tests/python/contrib/test_arm_compute_lib/test_add.py:118]: operations added to arm_compute_lib.rst
method renamed

**Review by lhutton1 — APPROVED** [review]
LGTM

**Comment by comaniac** [comment]
Thanks @d-smirnov @lhutton1
